### PR TITLE
Adding cudadecon

### PR DESCRIPTION
**Pull request by zbarry**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

Addresses https://github.com/tlambert03/pycudadecon/issues/15

### CODE REVIEW

File: recipes/cudadecon/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+  # If the installation is complex, or different between Unix and Windows, use separate bld.bat and build.sh files instead of this key.
+  # By default, the package will be built for the Python versions supported by conda-forge and for all major OSs.
+  # Add the line "skip: True  # [py<35]" (for example) to limit to Python 3.5 and newer, or "skip: True  # [not win]" to limit to Windows.
+  script: |
+    find .
+    mkdir cmake_build \
+      && cd cmake_build \
+      && cmake ${CMAKE_ARGS} -DCMAKE_BUILD_TYPE=Release ../src \
+      && make \
+      && make install
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:42]: Would suggest moving this into a `build.sh` to make it easier to follow

File: recipes/cudadecon/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,101 @@
+# Note: there are many handy hints in comments in this example -- remove them when you've finalized your recipe
+
+# Jinja variables help maintain the recipe as you'll update the version only here.
+# Using the name variable with the URL in line 14 is convenient
+# when copying and pasting from another recipe, but not really needed.
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:5]: All of these comments from the example recipe both here and below can be dropped

File: recipes/cudadecon/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    # Compilers are named 'c', 'cxx' and 'fortran'.
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}
+    - python
+    - cmake
+    - make  # [not win]
+    - ninja  # [win]
+  host:
+    - pip
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:56]: Need to include `python` in `host` for Python packages

```suggestion
    - python
    - pip
```

File: recipes/cudadecon/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    - boost-cpp
+    - libtiff
+    - fftw
+    # - ninja
+    # - cudatoolkit-dev =10.1.243
+    # - cmake
+    # - wget
+  run:
+    # - python
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:65]: Need `python` for Python packages

```suggestion
  run:
    - python
```

File: recipes/cudadecon/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      && make install
+
+requirements:
+  build:
+    # If your project compiles code (such as a C extension) then add the required compilers as separate entries here.
+    # Compilers are named 'c', 'cxx' and 'fortran'.
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}
+    - python
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:51]: No need to include `python` here. This is only used for build tools leveraged through their CLIs and not linked against (like `cmake`)

```suggestion
```

File: recipes/cudadecon/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    # - wget
+  run:
+    # - python
+
+test:
+  # Some packages might need a `test/commands` key to check CLI.
+  # List all the packages/modules that `run_test.py` imports.
+  imports:
+    - simplejson
+    - simplejson.tests
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:72]: Would replace these with appropriate `import` tests

File: recipes/cudadecon/yum_requirements.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+wget
+perl
```
Comment (jakirkham) [inline recipes/cudadecon/yum_requirements.txt:2]: Can these be included in `build` instead?

File: recipes/cudadecon/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    an optional C extension for a serious speed boost.
+  doc_url: https://simplejson.readthedocs.io/
+  dev_url: https://github.com/simplejson/simplejson
+
+extra:
+  recipe-maintainers:
+    # GitHub IDs for maintainers of the recipe.
+    # Always check with the people listed below if they are OK becoming maintainers of the recipe. (There will be spam!)
+    - LisaSimpson
+    - LandoCalrissian
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:101]: Would add your GH handle here

File: recipes/cudadecon/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  license_file: LICENSE.txt
+  summary: 'Simple, fast, extensible JSON encoder/decoder for Python'
+  # The remaining entries in this section are optional, but recommended.
+  description: |
+    simplejson is a simple, fast, complete, correct and extensible
+    JSON <https://json.org> encoder and decoder for Python 2.5+ and
+    Python 3.3+. It is pure Python code with no dependencies, but includes
+    an optional C extension for a serious speed boost.
+  doc_url: https://simplejson.readthedocs.io/
+  dev_url: https://github.com/simplejson/simplejson
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:94]: These need to be updated for the URLs, descriptions, etc. relevant to the package

**Review by jakirkham — COMMENTED** [review]
Thanks Zachary! 😄

Added some suggestions below. Please let us know if you have any questions 🙂

**Comment by zbarry** [comment]
Thanks, @jakirkham. All the metadata stuff I haven't gotten around to yet; I'm actually still struggling with compiling the code in the CI system right now - see [the build logs for the CUDA 10.2 node](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=298593&view=logs&jobId=232b7f5e-8d64-5438-fada-8478f379d226&j=232b7f5e-8d64-5438-fada-8478f379d226&t=e8516d7e-1153-5090-5807-53ae833fddd8) (the most relevant bit replicated below). 

Not sure if it's possible to change the version of gcc being used:

```
[ 20%] Building CXX object Buffers/CMakeFiles/Buffer.dir/PinnedCPUBuffer.cpp.o
[ 24%] Linking CXX static library ../linux-gnu-release/libBuffer.a
[ 24%] Built target Buffer
[ 28%] Building NVCC (Device) object CMakeFiles/libcudaDecon.dir/libcudaDecon_generated_geometryTransform.cu.o
In file included from /usr/local/cuda/include/cuda_runtime.h:83,
                 from <command-line>:
/usr/local/cuda/include/crt/host_config.h:138:2: error: #error -- unsupported GNU version! gcc versions later than 8 are not supported!
  138 | #error -- unsupported GNU version! gcc versions later than 8 are not supported!
      |  ^~~~~
CMake Error at libcudaDecon_generated_geometryTransform.cu.o.Release.cmake:221 (message):
  Error generating
  /home/conda/staged-recipes/build_artifacts/cudadecon_1617304563451/work/cmake_build/CMakeFiles/libcudaDecon.dir//./libcudaDecon_generated_geometryTransform.cu.o
```

It's kind of a weird error or my own misunderstanding of how this works, since I'm using conda-forge's supplied gcc and cuda compiler macros in the recipe, and the error here is happening in a CUDA header that I'm guessing is made available through the cuda macro.

I'm also not sure how the CI system works for compiling CUDA things, but since the node is explicitly named `linux_64_cuda_102`, does that mean it's only capable of building against cudatoolkit 10.2, as well? I think I would then have to go a different route than including `- {{ compiler('cuda') }}` in the recipe.

**Comment by jakirkham** [comment]
Maybe try adding a `conda_build_config.yaml` file in the recipe with the following?

```yaml
c_compiler_version:
  - 7
cxx_compiler_version:
  - 7
fortran_compiler_version:
  - 7
```

**Comment by jakirkham** [comment]
Ok looks like that resolved the compiler issue. Would take a look at the other suggestions above as well. These should help fix the remaining issues

**Comment by jakirkham** [comment]
@zbarry, did you have a chance to try incorporating the suggestions above? 🙂

**Comment by zbarry** [comment]
Hi @jakirkham - sorry about letting this stall for a while. Back on it now.

File: recipes/cudadecon/yum_requirements.txt (hunk 7 above)
Comment (zbarry) [reply recipes/cudadecon/yum_requirements.txt:2]: indeed they can

File: recipes/cudadecon/meta.yaml (hunk 4 above)
Comment (zbarry) [reply recipes/cudadecon/meta.yaml]: `pycudadecon` is actually a separate package that will call this `cudadecon` package (and I'll get around to that recipe next).

File: recipes/cudadecon/meta.yaml (hunk 3 above)
Comment (zbarry) [reply recipes/cudadecon/meta.yaml:27]: python not needed; see below

**Comment by zbarry** [comment]
Hey @tlambert03 would you mind leaving a comment here confirming you would like to be a maintainer on this? Thanks!

Also: for Windows CUDA support and other versions of CUDA on Linux, I'm told how this will go is in a day or two, the build system will automatically issue PRs for them that we can then merge.

**Comment by tlambert03** [comment]
> Hey @tlambert03 would you mind leaving a comment here confirming you would like to be a maintainer on this? Thanks!

yes, please include me as a maintainer here.  thanks again!

**Comment by zbarry** [comment]
@conda-forge/help-c-cpp this is now ready for review. Thanks!

File: recipes/cudadecon/yum_requirements.txt (hunk 7 above)
Comment (jakirkham) [inline recipes/cudadecon/yum_requirements.txt:2]: Since these are in `build` now, can this file be dropped?

File: recipes/cudadecon/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  skip: True  # [win or osx or (cuda_compiler_version in ('None', 'undefined'))]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - {{ compiler('cuda') }}
+    - cmake
+    - make  # [not win]
+    - ninja  # [win]
```
Comment (jakirkham) [inline recipes/cudadecon/meta.yaml:23]: If we are building on Windows, should we include a `bld.bat` or are we skipping that (in which case maybe `ninja` and the selector above can be dropped)?

Comment (zbarry) [reply recipes/cudadecon/meta.yaml:23]: @isuruf mentioned in the gitter chat that we will be getting an automated PR in a couple days following the creation of the feedstock which will help us build the Windows version. Was just leaving `ninja` in there in anticipation of that, though if you would like it removed until that time, can do that as well.

File: recipes/cudadecon/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  host:
+    - boost-cpp
+    - libtiff
+    - fftw
+  run:
+    - boost-cpp
+    - _openmp_mutex
```
Comment (isuruf) [inline recipes/cudadecon/meta.yaml:30]: This is wrong. See our docs on openmp

Comment (zbarry) [reply recipes/cudadecon/meta.yaml:30]: That was actually to satisfy an error message that was roughly something like "possible overlinking or runtime dependency not present". It's not clear from https://conda-forge.org/docs/maintainer/knowledge_base.html#openmp that adding `libgomp` to the build deps does anything more than making something already implicitly included explicitly included instead (in which case, I'd still get that error). Maybe I'm missing something else? Will go ahead and try it either way.

Comment (zbarry) [reply recipes/cudadecon/meta.yaml]: Annnnd it worked :)

File: recipes/cudadecon/meta.yaml (hunk 10 above)
Comment (zbarry) [reply recipes/cudadecon/meta.yaml]: Removing for now

**Comment by zbarry** [comment]
Thanks all for your reviews! Anything else to address? :)

**Review by jakirkham — APPROVED** [review]
LGTM. Thanks @zbarry 😄

**Review by tlambert03 — APPROVED** [review]
great work, thank you!

**Comment by jakirkham** [comment]
Thanks all! 😄

**Comment by jakirkham** [comment]
Will there be a `pycudadecon` PR in the near future? 😉

(no pressure. merely curious 🙂)

**Comment by tlambert03** [comment]
definitely! :) I'm very excited to have this in for exactly that reason (thank you again so much for pitching in)

not sure when I'll get to it, but I think "near future" is safe to say

**Pull request merged**